prometheus/go_collector_go117.go

```diff
@@ -303,37 +317,38 @@ type batchHistogram struct {
 	sum     float64 // Used if hasSum is true.
 }
 
-func newBatchHistogram(desc *Desc, hasSum bool) *batchHistogram {
-	h := &batchHistogram{desc: desc, hasSum: hasSum}
+func newBatchHistogram(desc *Desc, buckets []float64, hasSum bool) *batchHistogram {
+	h := &batchHistogram{
+		desc:    desc,
+		buckets: buckets,
+		counts:  make([]uint64, len(buckets)-1),
+		hasSum:  hasSum,
+	}
 	h.init(h)
 	return h
 }
 
 // update updates the batchHistogram from a runtime/metrics histogram.
 //
 // sum must be provided if the batchHistogram was created to have an exact sum.
+// h.buckets must be a strict subset of his.Buckets.
 func (h *batchHistogram) update(his *metrics.Float64Histogram, sum float64) {
 	counts, buckets := his.Counts, his.Buckets
-	// Skip a -Inf bucket altogether. It's not clear how to represent that.
-	if math.IsInf(buckets[0], -1) {
-		buckets = buckets[1:]
-		counts = counts[1:]
-	}
 
 	h.mu.Lock()
 	defer h.mu.Unlock()
 
-	// Check if we're initialized.
-	if h.buckets == nil {
-		// Make copies of counts and buckets. It's really important
-		// that we don't retain his.Counts or his.Buckets anywhere since
-		// it's going to get reused.
-		h.buckets = make([]float64, len(buckets))
-		copy(h.buckets, buckets)
-
-		h.counts = make([]uint64, len(counts))
+	// Reduce buckets.
+	var j int
+	var total uint64
+	for i, count := range counts {
+		total += count
```

Comment:
nit: Hm, I feel if we get rid `total` variable and add directly in `h.buckets` it would be more clear? Fewer code lines, and to me `total` purpose was a bit misleading (it's not total, it's actually temp total between buckets). Not a blocker (:

Reply:
done, but I broke out the clearing separately. I think it's better this way.